lib/ts/hugepages.h

```diff
@@ -20,13 +20,26 @@
  */
 #ifndef _hugepages_h_
 #define _hugepages_h_
+#include "ink_memory.h"
 
 #include <cstring>
 
+#define HUGEPAGE_SIZE  (PAGE_SIZE<<9)
+
 size_t ats_hugepage_size(void);
 bool ats_hugepage_enabled(void);
 void ats_hugepage_init(int);
 void *ats_alloc_hugepage(size_t);
 bool ats_free_hugepage(void *, size_t);
+void *ats_alloc_hugepage_stack(size_t);
+
+void *huge_normal_alloc(void *chunk, size_t size, size_t alignment, bool *zero, bool *commit, unsigned arena_ind);
+void *huge_nodump_alloc(void *chunk, size_t size, size_t alignment, bool *zero, bool *commit, unsigned arena_ind);
+bool huge_dalloc(void *chunk, size_t size, bool committed, unsigned arena_ind);
+bool huge_commit(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind);
+bool huge_decommit(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind);
+bool huge_purge(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind);
+bool huge_split(void *chunk, size_t size, size_t size_a, size_t size_b, bool committed, unsigned arena_ind);
+bool huge_merge(void *chunk_a, size_t size_a, void *chunk_b, size_t size_b, bool committed, unsigned arena_ind);
```

Comment:
Creating a block of 2Meg (512 pages) ... is instant and as fast as mmap can do it.   It never will fail unless memory is gone.   There is no wait and there is no muss and no fuss and it is transparent.

Why do we want hugepages, then?   Not because we demand them (for some hardware).   We need memory *simplified* / *copied* / *rearranged* down so that contiguous pages are put directly into a single hugepage ... over time.   Best of all would be simply for it to happen in the background.

That's what hugepaged does.   

Because of that ... everything in these hooks is *aligned* and *allocated* in hugepage-sized groups (512 aligned correctly).   This system then lets jemalloc still demand changes to segments and offsets within the page.   That means commit/decommit are often no-ops that simply return with no effect: a single hugepage is already present (or may be) and decommiting part of it is somewhat silly.

De-allocating an aligned set of hugepages ... is performed as needed.   However arenas are rarely shut down entirely ... to that degree.  A group of 512 pages together will still need *some* page or another left behind.